### ISSUE TITLE
Bounds checking

### ISSUE THREAD

**Issue by anntzer** [inducer/cnd#1]
This is a _very_ interesting approach.
It seems it would be useful to allow some form of bounds-checking by expanding `a[i, j]` to `(assert(0 <= i && i < m && 0 <= j && j < n), a[i * n + j])`.

**Comment by inducer** [comment]
Agreed, that would be nice to have. It wouldn't require a whole lot of work. One possible issue is that `assert` isn't an expression, but that's easy enough to work around.

I'd be happy to take a patch.
